smartsim/_core/mli/workermanager.py

```diff
@@ -0,0 +1,976 @@
+        ] = {}
+        """a collection of workers the manager is controlling"""
+        self._upstream_queue: t.Optional[mp.Queue[bytes]] = None
+        """the queue the manager monitors for new tasks"""
+        self._feature_store: FeatureStore = feature_store
+        """a feature store to retrieve models from"""
+        self._worker = worker
+        """The ML Worker implementation"""
+
+    @property
+    def upstream_queue(self) -> "t.Optional[mp.Queue[bytes]]":
+        """Return the queue used by the worker manager to receive new work"""
+        return self._upstream_queue
+
+    @upstream_queue.setter
+    def upstream_queue(self, value: "mp.Queue[bytes]") -> None:
+        """Set/update the queue used by the worker manager to receive new work"""
+        self._upstream_queue = value
+
+    def _on_iteration(self, timestamp: int) -> None:
+        """Executes calls to the machine learning worker implementation to complete
+        the inference pipeline"""
+        logger.debug(f"{timestamp} executing worker manager pipeline")
+
+        if self.upstream_queue is None:
+            logger.warning("No queue to check for tasks")
+            return
+
+        msg: bytes = self.upstream_queue.get()
+        request = self._worker.deserialize(msg)
+        fetch_model_result = self._worker.fetch_model(request, self._feature_store)
+        model_result = self._worker.load_model(request, fetch_model_result)
+        fetch_input_result = self._worker.fetch_inputs(
+            request,
+            self._feature_store,
+        )  # we don't know if they'lll fetch in some weird way
+        # they potentially need access to custom attributes
+        # we don't know what the response really is... i have it as bytes
+        # but we just want to advertise that the contract states "the output
+        # will be the input to transform_input... "
+
+        transform_result = self._worker.transform_input(request, fetch_input_result)
+
+        # batch: t.Collection[_Datum] = transform_result.transformed_input
+        # if self._batch_size:
+        #     batch = self._worker.batch_requests(transform_result, self._batch_size)
+
+        # todo: what do we return here? tensors? Datum? bytes?
+        results = self._worker.execute(request, model_result, transform_result)
+
+        # todo: inference reply _must_ be replaced with the mli schemas reply
+        reply = InferenceReply()
+
+        # only place into feature store if keys are provided
+        if request.output_keys:
+            output_keys = self._worker.place_output(
+                request, results, self._feature_store
+            )
+            reply.output_keys = output_keys
+        else:
+            reply.outputs = results.predictions
+
+        serialized_output = self._worker.serialize_reply(reply)
+
+        callback_channel = request.callback
+        if callback_channel:
+            callback_channel.send(serialized_output)
+
+    def _can_shutdown(self) -> bool:
```

Comment:
Maybe, I missed it, but was ``self._worker`` ever added to ``self._workers``?  If not, do we still want ``self._workers`` at this point?  I'd be happy to break out multiple workers as a separate ticket after we do batching

Reply:
Added 717 for follow-up work